README.md
![Figure 1: The stages and the timing of the experiment.](https://raw.github.com/jozefdobos/gnret/master/docs/images/fig1.png)
<b>Figure 1:</b> The stages and the timing of the experiment.

# 1 Introduction
Currently at [version 0.7.4](https://github.com/jozefdobos/gnret/blob/master/CHANGELOG.md), the Gender-Neutral Real-Effort Task (GNRET) is a collection of individual feedback driven end-user tasks and a master control application to monitor the ongoing experiment in real-time from a remote location. Under the supervision of [Dr. Victoria Prowse](https://faculty.cit.cornell.edu/vlp33/), this study was designed by Zdenka Kissová as a partial fullfilment of the requirements for the degree of M.Phil. in [Economics](http://www.economics.ox.ac.uk/) at [University of Oxford](http://www.ox.ac.uk/). The experiments were conducted in the experimental laboratory of the [Nuffield Centre for Experimental Social Sciences (CESS)](https://cess-nuffield.nuff.ox.ac.uk/) at [University of Oxford](http://www.ox.ac.uk/). The actual software was delivered by [Jozef Doboš](https://github.com/jozefdobos) from the [Department of Computer Science](http://www.cs.ucl.ac.uk/), [University College London](http://www.ucl.ac.uk/).

# 2 Program Flow
Figure 1 outlines the stages and timing of the experiment. Each experimental session lasts for approximately 90 minutes and consists of seven (one practice and six paying) 6-minute rounds. In each 6-minute round, the initial 3 minutes are allocated to the verbal component of the task (word spotting), followed by another 3 minutes of numerical questions (adding up) all of which are automatically timed out. There is a 4-minute break after each round. During the first minute of this break, participants can always see their individual total points score (the sum of the verbal and numerical point scores) for the respective round on their screens. In the last three minutes of the break, Treatment Group subjects receive treatment-specific relative performance feedback. The demographics and behavioural data questionnaire is displayed after the last experimental round. The questionnaire asks about basic demographics such as gender, age, country of origin and degree of study as well as the subject's perceived level of competitiveness, familiarity with the task, satisfaction with the performance and the potential use of performance goals.


<img src="https://raw.github.com/jozefdobos/gnret/master/docs/images/fig2a.png" alt="Figure 2: (a) The letter grid for word spotting used in the Practice Round across all sessions." width=430>&nbsp;<img src="https://raw.github.com/jozefdobos/gnret/master/docs/images/fig2b.png" alt="Figure 2: (b) The numerical grid with an adding up question used in the Practice Round across all sessions." width=430>

<b>Figure 2:</b> (a) The letter grid for word spotting used in the Practice Round across all sessions. (b) The numerical grid with an adding up question used in the Practice Round across all sessions.

## 2.1 Word Spotting Task
During the word spotting, each subject is presented with a 15 by 15 grid of capital letters, see Figure 2(a), and is asked to look for English words in horizontal, vertical or diagonal direction, forward or backward. A word is selected using the mouse (This ensures that the subjects are not disadvantaged based on their typing skills.). The subjects need to place the cursor of the mouse over the starting letter of a proposed word, pull the cursor to the last letter of that word and then release the cursor (The first and last letter recognition prevents pure guessing.). If the selected word is valid, it remains highlighted and a point is awarded. No points are awarded or subtracted for incorrect entries so as to avoid potential hostile framing or punishment effects in the subjects' behaviour.

## 2.2 Numerical Task
After the initial three minutes elapse, an adding up question for a pair of 2-digit numbers automatically appear on the screen. The subject needs to correctly sum the numbers to win a point and proceed to the next numerical question. Calculators, pens and paper are not allowed. The answer is entered using the mouse and a calculator-style number pad on the screen. As in the verbal component of the task, no points are awarded or subtracted for incorrect entries. The points from the verbal and numerical task components in a specic round are added up to form the total point score for that round. Identical letter grids and number questions in the same order are used across all sessions and hence all treatment groups.


<img src="https://raw.github.com/jozefdobos/gnret/master/docs/images/fig3a.png" alt="Figure 3: Individual and relative performance feedback as displayed during (a) the first and (b) the last three minutes of the break to Treatment Group 1. Note that the first-minute individual performance feedback was displayed in the same manner to all subjects across the experiment." width=430>&nbsp;<img src="https://raw.github.com/jozefdobos/gnret/master/docs/images/fig3b.png" alt="Figure 3: Individual and relative performance feedback as displayed during (a) the first and (b) the last three minutes of the break to Treatment Group 1. Note that the first-minute individual performance feedback was displayed in the same manner to all subjects across the experiment." width=430>

<b>Figure 3:</b> Individual and relative performance feedback as displayed during (a) the first and (b) the last three minutes of the break to Treatment Group 1. Note that the first-minute individual performance feedback was displayed in the same manner to all subjects across the experiment.

## 2.3 Performance Feedback
Subjects are divided into four groups with the same selection and order of the tasks, but with different performance feedback as follows:
- Control Group (C) - no relative performance feedback.
- Treatment Group 1 (T1) - computerised private relative performance feedback.
- Treatment Group 2 (T2) - personal private relative performance feedback.
- Treatment Group 3 (T3) - personal public relative performance feedback.

<img src="https://raw.github.com/jozefdobos/gnret/master/docs/images/fig4.png" alt="Figure 4: FTP Login window showing location of the log file." width=250>

<b>Figure 4:</b> FTP Login window showing location of the log file.

# 3 Technical Information
The front-end Graphical User Interface (GUI) is written in Java Swing with networking provided by the Apache Commons library executed in an exclusive full-screen mode to avoid any unnecessary distractions.

Once started, each client application, the one that the subjects interact with, opens up a File Transfer Protocol (FTP) connection to the master server from which it polls for a start and every round signals triggered by the admin interface. Only if the corresponding buttons are set, see Figure 5 for toolbar options, will the individual sections as outlined in Figure 1 automatically proceed. Client saves locally a text le with all so far collected responses which is uploaded to the main FTP server during each break. Admin program automatically calculates the rank of each client and saves the results to the FTP server from which the client if in T1 mode requests this information and displays it on the screen. Intermediate results as well as the final questionnaire answers are viewable in the admin interface.

## 3.1 Error Handling
In the case of a crash, there is always a local copy of the subject's responses as well as a program log file. If the FTP connection is dropped, the client will automatically attempt reconnection with a randomised interval of 30 to 60 seconds in between as to prevent overloading the server.

## 3.2 Requirements
- Java Runtime Environment v1.4+.
- Java Webstart plugin.
- MS Windows or Linux based FTP server with password protected write access. 
 
Downloadable from http://www.java.com/en/download/index.jsp

![Figure 5: FTP Login window showing location of the log file.](https://raw.github.com/jozefdobos/gnret/master/docs/images/fig5.png)

<b>Figure 5:</b> Admin interface for live monitoring of the experiment.

## 3.3 External Dependencies
<dl><dt>Apache Commons Net 2.2.</dt><dd>This standard Apache release library provides networking functionality through FTP and socket connection wrappers and is distributed under the Apache License Version 2.0 (January 2004), can be found at: http://www.apache.org/
licenses/.</dd>
<dt>Java File Graphics 1.0.</dt><dd> Former Sun Microsystems created a standard icon graphics library
to distribute with their Java Swing applications.</dd> 
<dt>Jazzy Core Spell Checker.<dt><dd>Jazzy Core implements spell checking algorithm similar to as
pell and is distributed under GNU Lesser General Public License Version 2.1 (February 1999). The source and precompiled libraries can be found at: http://sourceforge.net/projects/jazzy. </dd>
<dt>ISpell Word Lists.</dt><dd>A word list compiled from American and British English spelling lists of the Ispell Version 3.1.20 distributed under a royalty free license, a copyright 1993 of Geoff Kuenning, Granada Hills, CA, USA. The individual dictionaries can be found at: http://wordlist.sourceforge.net/.</dd>
</dl>


CHANGELOG.md
# Change Log

### v0.7.3 (17/02/2011)
- Removed rank from T1 practice progress waiting.

### v0.7.2 (15/02/2011)
- Removed rank from T1 practice round.

### v0.7.1 (11/02/2011)
- Added more reconnection attempts in hope of being able to recover from socket write error.
- Added Local passive mode tick box in FTP Login Window.
- Set data timeout to 2.5 min in Admin Window.

### v0.7 (10/02/2011)
- BugFix: Removed timeout on the Questionnaire.
- Added reconnection request modal dialog in Admin Window when exporting data.
- Increased minimal JVM heap size to 128m.
- JVM heap size printed to console.

### v0.6.9 (10/02/2011)
- Added fields in Questionnaire for age, field of study and drop downs for countries. There is also option to specify into a text field if suitable selection is not present.

### v0.6.8 (07/02/2011)
- Tried to fix T1 rank delivery.
- Each new connected computer has now unique ID based on previous highest ID on the server +1.
- Console messages are now logged into gnret.log file under ~\My Documents\gnret\gnret.log
- BugFix: Fixed export of data table to CSV (commas now sanitised).

### v0.6.7 (06/02/2011)
- Finalised for Monday (07/02/2011) testing.
- Added cess_gnret.jnlp file.

### v0.6.6 (06/02/2011)
- BugFix: Missing NOOP in Experiment Window.

### v0.6.5 (06/02/2011)
- Added select all files functionality.
- On refresh, gnret folder is automatically expanded if present.
- BugFix: Missing column 6 in rank calculation of Experiment Window.
- BugFix: Other comments text field did not propagate answers in the Questionnaire.
- BugFix: Rank display in T1 group. Now rank calculation happens in a serial call after results uploaded to the server.
- BugFix: There used to be a memory leak of loading dictionary 7 times.

### v0.6.4 (05/02/2011)
- All catch statements in Experiment window now log into console of FTP window.
- Added delete files popupmenu in the filetree as well as toolbar and manu bar of Amin window.
- Modified reconnection attempts within Experiment window.
- Added 5 min connection timeout

### v0.6.3 (03/02/2011)
- All error messages now bubble up to FTP Login window.
- Rank tiebreak resolved randomly based on current round as seed number.

### v0.6.2 (03/02/2011)
- Added support for going forwards (Shift + Right Arrow Key) and backwards (Shift + Left Arrow Key) in the experiment window.
- Modified textual labels within experiment as requested
- BugFix: When long delay from starting up the experiment window to actual pressing of the start button happened, window got disconnected experiencing socket connection error from which it could not recover. Fixed by forcing disconnection after window appears on the screen and reconnecting on button press.
- BugFix: If file upload error in experiment window occurred, it used to pass error message as file name on retry.
- BugFix: Experiment window did not set time hence no NOOP messages were sent.

### v0.6.1 (03/02/2011)
- Changed visual appearance of Questionnaire
- Added support for questionnaire in the Admin window

### v0.6 (02/02/2011)
- Redesigned the screen switching algorithm
- Added automatic rank delivery for T1 group
- Added randomised NOOP messages
- Added automatic TCP/IP port selection when field left blank in FTP Login window
- BugFix: Fixed timing delay for progress check
- BugFix: Fixed message on waiting progress bar in Admin window
- BugFix: Fixed double timer speed (time used to progress faster than desired)
- BugFix: Reset cursor when selecting About dialog from Admin window menu bar

### v0.5 (01/02/2011)
- Transformed into multithreaded application to make GUI more responsive
- Added progressive file loading in Admin window
- Added hand made letter grids for all 7 (6+1) rounds
- Added improved dictionary with manually prunned 2 and 3 letter words
- Added 'Start experiment' button in Admin window
- Added automatic RANK calculation
- Added Start/Stop experiment functionality for each round separately

### v0.4 (27/01/2011)
- Added table printing functionality
- Refresh and Reconnect functionality in Admin window implemented
- Added Admin window



src/eu/dobos/jozef/gnret/gui/AbstractGNRETFrame.java
/**
*  Copyright (C) 2011 Jozef Dobos
*
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU Affero General Public License as
*  published by the Free Software Foundation, either version 3 of the
*  License, or (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU Affero General Public License for more details.
*
*  You should have received a copy of the GNU Affero General Public License
*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package eu.dobos.jozef.gnret.gui;

import java.io.IOException;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.util.Random;

import javax.swing.JFrame;
import javax.swing.SwingWorker;

import org.apache.commons.net.ftp.FTPConnectionClosedException;

import eu.dobos.jozef.gnret.communication.JakartaFtpWrapper;
import eu.dobos.jozef.gnret.filehandling.FileWritingHandler;
import eu.dobos.jozef.gnret.gui.utils.Settings;
import eu.dobos.jozef.gnret.gui.utils.Utils;
import eu.dobos.jozef.gnret.time.TimeKeeper;

public abstract class AbstractGNRETFrame extends JFrame {

	/**
	 * 
	 */
	private static final long serialVersionUID = 727146086756011811L;

	protected TimeKeeper timeKeeper;
	protected int roundCounter = 0;
	private Random randomizer = new Random();
	protected long nextNoopTimeCheck = getNextNoopCheckTime();
	protected long previousTime = 0;

	public static final int CONNECTION_SCREEN = 0;
	public static final int PRACTICE_ROUND_GRID_SCREEN = 1;
	public static final int PRACTICE_ROUND_NUMERICAL_SCREEN = 2;
	public static final int PRACTICE_ROUND_RESULTS_SCREEN = 3;
	public static final int PROGRESS_CHECK_PRACTICE_ROUND = 4;

	public static final int PAYING_ROUND_ONE_GRID_SCREEN = 5;
	public static final int PAYING_ROUND_ONE_NUMERICAL_SCREEN = 6;
	public static final int PAYING_ROUND_ONE_RESULTS_SCREEN = 7;
	public static final int PROGRESS_CHECK_PAYING_ROUND_ONE = 8;

	public static final int PAYING_ROUND_TWO_GRID_SCREEN = 9;
	public static final int PAYING_ROUND_TWO_NUMERICAL_SCREEN = 10;
	public static final int PAYING_ROUND_TWO_RESULTS_SCREEN = 11;
	public static final int PROGRESS_CHECK_PAYING_ROUND_TWO = 12;

	public static final int PAYING_ROUND_THREE_GRID_SCREEN = 13;
	public static final int PAYING_ROUND_THREE_NUMERICAL_SCREEN = 14;
	public static final int PAYING_ROUND_THREE_RESULTS_SCREEN = 15;
	public static final int PROGRESS_CHECK_PAYING_ROUND_THREE = 16;

	public static final int PAYING_ROUND_FOUR_GRID_SCREEN = 17;
	public static final int PAYING_ROUND_FOUR_NUMERICAL_SCREEN = 18;
	public static final int PAYING_ROUND_FOUR_RESULTS_SCREEN = 19;
	public static final int PROGRESS_CHECK_PAYING_ROUND_FOUR = 20;

	public static final int PAYING_ROUND_FIVE_GRID_SCREEN = 21;
	public static final int PAYING_ROUND_FIVE_NUMERICAL_SCREEN = 22;
	public static final int PAYING_ROUND_FIVE_RESULTS_SCREEN = 23;
	public static final int PROGRESS_CHECK_PAYING_ROUND_FIVE = 24;

	public static final int PAYING_ROUND_SIX_GRID_SCREEN = 25;
	public static final int PAYING_ROUND_SIX_NUMERICAL_SCREEN = 26;
	public static final int PAYING_ROUND_SIX_RESULTS_SCREEN = 27;
	public static final int PROGRESS_CHECK_PAYING_ROUND_SIX = 28;

	public static final int QUESTIONNAIRE_SCREEN = 29;

	public static final String[] EXPERIMENT_DESCRIPTION = { "Start screen",
			Settings.PRACTICE_ROUND + " GRID",
			Settings.PRACTICE_ROUND + " NUMERICAL",
			Settings.PRACTICE_ROUND + " RESULTS", "",
			Settings.PAYING_ROUND + " 1 GRID",
			Settings.PAYING_ROUND + " 1 NUMERICAL",
			Settings.PAYING_ROUND + " 1 RESULTS", "",
			Settings.PAYING_ROUND + " 2 GRID",
			Settings.PAYING_ROUND + " 2 NUMERICAL",
			Settings.PAYING_ROUND + " 2 RESULTS", "",
			Settings.PAYING_ROUND + " 3 GRID",
			Settings.PAYING_ROUND + " 3 NUMERICAL",
			Settings.PAYING_ROUND + " 3 RESULTS", "",
			Settings.PAYING_ROUND + " 4 GRID",
			Settings.PAYING_ROUND + " 4 NUMERICAL",
			Settings.PAYING_ROUND + " 4 RESULTS", "",
			Settings.PAYING_ROUND + " 5 GRID",
			Settings.PAYING_ROUND + " 5 NUMERICAL",
			Settings.PAYING_ROUND + " 5 RESULTS", "",
			Settings.PAYING_ROUND + " 6 GRID",
			Settings.PAYING_ROUND + " 6 NUMERICAL",
			Settings.PAYING_ROUND + " 6 RESULTS", "" };

	public static final long[] _EXPERIMENT_TIMING = {
		Settings.PROCEEDING_CHECK_DURATION, // start session check
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			Settings.PROCEEDING_CHECK_DURATION,// practice

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 1
			Settings.PROCEEDING_CHECK_DURATION,

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 2
			Settings.PROCEEDING_CHECK_DURATION,

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 3
			Settings.PROCEEDING_CHECK_DURATION,

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 4
			Settings.PROCEEDING_CHECK_DURATION,

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 5
			Settings.PROCEEDING_CHECK_DURATION,

			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS,
			TimeKeeper.SIX_SECONDS, // round 6
			Settings.PROCEEDING_CHECK_DURATION,
			Settings.PROCEEDING_CHECK_DURATION // Questionnaire keep alive
	};

	public static final long[] EXPERIMENT_TIMING = {
			Settings.PROCEEDING_CHECK_DURATION, // start session check
			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // practice
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 1
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 2
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 3
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 4
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION,
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 5
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.TASK_DURATION, 
			Settings.TASK_DURATION,
			Settings.RESULT_DURATION, // round 6
			Settings.PROCEEDING_CHECK_DURATION,

			Settings.PROCEEDING_CHECK_DURATION // Questionnaire keep alive
	};
	
	public AbstractGNRETFrame(String name) {
		super(name);
	}

	// abstract
	public abstract void consoleAppendText(String text);
	protected abstract void returnTime(long time);
	protected abstract void switchScreen();


	public static boolean isProgressCheckScreen(int screenCounter) {
		return (screenCounter == CONNECTION_SCREEN
				|| screenCounter == PROGRESS_CHECK_PRACTICE_ROUND
				|| screenCounter == PROGRESS_CHECK_PAYING_ROUND_ONE
				|| screenCounter == PROGRESS_CHECK_PAYING_ROUND_TWO
				|| screenCounter == PROGRESS_CHECK_PAYING_ROUND_THREE
				|| screenCounter == PROGRESS_CHECK_PAYING_ROUND_FOUR
				|| screenCounter == PROGRESS_CHECK_PAYING_ROUND_FIVE || screenCounter == PROGRESS_CHECK_PAYING_ROUND_SIX);
	}

	public static boolean isResultsScreen(int screenCounter) {
		return (screenCounter == PRACTICE_ROUND_RESULTS_SCREEN
				|| screenCounter == PAYING_ROUND_ONE_RESULTS_SCREEN
				|| screenCounter == PAYING_ROUND_TWO_RESULTS_SCREEN
				|| screenCounter == PAYING_ROUND_THREE_RESULTS_SCREEN
				|| screenCounter == PAYING_ROUND_FOUR_RESULTS_SCREEN
				|| screenCounter == PAYING_ROUND_FIVE_RESULTS_SCREEN || screenCounter == PAYING_ROUND_SIX_RESULTS_SCREEN);
	}

	public static void saveTextToLocalFile(AbstractGNRETFrame frame,
			String fileName, String text) {
		try {
			FileWritingHandler.appendToFile(fileName, text, true);
		} catch (IOException e) {
			frame.consoleAppendText("E10: " + e.getMessage());
			String titleDescription = "Error while saving to a local file";
			if (0 == Utils.showRetryCancelErrorDialog(frame, e.getMessage(),
					titleDescription)) {
				saveTextToLocalFile(frame, fileName, text);
			}
		}
	}

	/***
	 * 
	 * @param checkStart
	 *            true for start experiment, false for individual rounds
	 * @return
	 */
	public static void switchScreenIfCanProgress(final boolean checkStart,
			final JakartaFtpWrapper ftp, final AbstractGNRETFrame frame,
			final int roundCounter, final int screenCounter) {		
		SwingWorker<boolean[], Void> worker = new SwingWorker<boolean[], Void>() {
			@Override
			protected boolean[] doInBackground() throws Exception {
				boolean[] ret = { false };
				try {
					frame.consoleAppendText("Checking if can proceed... round counter: " + roundCounter);
					if (!ftp.isConnected()) {
						frame.consoleAppendText("Proceed check -- Disconnected, trying to reconnect");

						tryToReconnectAndLog(ftp, true, frame);
					}
					if (checkStart) {
						ret[0] = JakartaFtpWrapper.canStart(ftp);
					} else {
						ret[0] = JakartaFtpWrapper
								.canProceed(ftp, roundCounter);
					}
				} catch (Exception e) {
					frame.consoleAppendText("E7: " + e.getMessage());
					tryToReconnectAndLog(ftp, true, frame);
				}
				return ret;
			}

			@Override
			public void done() {
				try {
					boolean[] canProceed = get();

					if (canProceed != null) {
						if (canProceed[0]) {
							frame.consoleAppendText("Result: canProceed is true");
							if (frame.timeKeeper != null) {
								frame.timeKeeper.stop();
								frame.timeKeeper = null;
							}
							if (!(screenCounter == CONNECTION_SCREEN)) {
								frame.incrementRoundCounter();
							}
							frame.switchScreen();
						} else {
							frame.consoleAppendText("Result: canProceed is false");
						}
					}
				} catch (InterruptedException ignore) {
					frame.consoleAppendText("E8: " + ignore.getMessage());
				} catch (java.util.concurrent.ExecutionException e) {
					String why = null;
					Throwable cause = e.getCause();
					if (cause != null) {
						why = cause.getMessage();
					} else {
						why = e.getMessage();
					}
					frame.consoleAppendText("E9: " + why);
				}
			}
		};
		worker.execute();
	}

	public static boolean tryToReconnectAndLog(JakartaFtpWrapper ftp,
			boolean changeToDefaultDir, AbstractGNRETFrame frame)
			throws UnknownHostException, FTPConnectionClosedException,
			SocketException, IOException {
		boolean ret = false;
		if (ret = ftp.tryToReconnectToFTP(changeToDefaultDir)) {
			frame.consoleAppendText("Reconnected: " + ftp.getReplyString());
		} else
			frame.consoleAppendText("Reconnect replied: "
					+ ftp.getReplyString());
		return ret;
	}

	public synchronized void decrementRoundCounter() {
		roundCounter--;
	}

	/***
	 * Returns random number between 30 and 60 seconds inclusive
	 * 
	 * @return
	 */
	public long getNextNoopCheckTime() {
		return TimeKeeper.getTimeInMilliseconds(0, 30 + randomizer.nextInt(61));
	}

	public int getRoundCounter() {
		return roundCounter;
	}

	// Fast forward testing time
	public synchronized void incrementRoundCounter() {
		roundCounter++;
	}

	public synchronized void setRoundCounter(int round) {
		roundCounter = round;
	}

	public void uploadLocalFileToServer(final AbstractGNRETFrame frame,
			final JakartaFtpWrapper ftp, final String fileName) {
		try {
			if (!ftp.isConnected()) {
				consoleAppendText("File Upload -- Disconnected, trying to reconnect");
				tryToReconnectAndLog(ftp, true, frame);
			}
			consoleAppendText("Trying to upload file: " + fileName
					+ " to the server");
			boolean uploaded = ftp.uploadFile(
					FileWritingHandler.getAbsolutePathToLocalFile(fileName),
					fileName);
			if (uploaded)
				consoleAppendText("Uploaded successfully");
		} catch (FTPConnectionClosedException ftpe) {
			frame.consoleAppendText("E11: " + ftpe.getMessage());
			try {
				ftp.disconnect();
			} catch (Exception e2) {
				frame.consoleAppendText("E12: " + e2.getMessage());
			}
			try {
				tryToReconnectAndLog(ftp, true, frame);
			} catch (Exception e3) {
				frame.consoleAppendText("E13: " + e3.getMessage());
			}
		} catch (Exception e) {
			String titleDescription = "Error while uploading data to the server";
			frame.consoleAppendText(titleDescription);
			frame.consoleAppendText("E14: " + e.getMessage());
			if (0 == Utils.showRetryCancelErrorDialog(frame, e.getMessage(),
					titleDescription)) {
				try {
					tryToReconnectAndLog(ftp, true, frame);
				} catch (Exception e4) {
					frame.consoleAppendText("E16: " + e4.getMessage());
				}
				uploadLocalFileToServer(frame, ftp, fileName);
			}
		}
	}
}
